humanoid_league_team_communication/humanoid_league_team_communication/humanoid_league_team_communication.py

```diff
@@ -1,438 +1,229 @@
-        self.role_mapping = (
-            (robocup_extension_pb2.Role.ROLE_UNDEFINED, Strategy.ROLE_UNDEFINED),
-            (robocup_extension_pb2.Role.ROLE_IDLING, Strategy.ROLE_IDLING),
-            (robocup_extension_pb2.Role.ROLE_OTHER, Strategy.ROLE_OTHER),
-            (robocup_extension_pb2.Role.ROLE_STRIKER, Strategy.ROLE_STRIKER),
-            (robocup_extension_pb2.Role.ROLE_SUPPORTER, Strategy.ROLE_SUPPORTER),
-            (robocup_extension_pb2.Role.ROLE_DEFENDER, Strategy.ROLE_DEFENDER),
-            (robocup_extension_pb2.Role.ROLE_GOALIE, Strategy.ROLE_GOALIE),
-        )
-        self.action_mapping = (
-            (robocup_extension_pb2.Action.ACTION_UNDEFINED, Strategy.ACTION_UNDEFINED),
-            (robocup_extension_pb2.Action.ACTION_POSITIONING, Strategy.ACTION_POSITIONING),
-            (robocup_extension_pb2.Action.ACTION_GOING_TO_BALL, Strategy.ACTION_GOING_TO_BALL),
-            (robocup_extension_pb2.Action.ACTION_TRYING_TO_SCORE, Strategy.ACTION_TRYING_TO_SCORE),
-            (robocup_extension_pb2.Action.ACTION_WAITING, Strategy.ACTION_WAITING),
-            (robocup_extension_pb2.Action.ACTION_KICKING, Strategy.ACTION_KICKING),
-            (robocup_extension_pb2.Action.ACTION_SEARCHING, Strategy.ACTION_SEARCHING),
-            (robocup_extension_pb2.Action.ACTION_LOCALIZING, Strategy.ACTION_LOCALIZING),
-        )
-        self.side_mapping = (
-            (robocup_extension_pb2.OffensiveSide.SIDE_UNDEFINED, Strategy.SIDE_UNDEFINED),
-            (robocup_extension_pb2.OffensiveSide.SIDE_LEFT, Strategy.SIDE_LEFT),
-            (robocup_extension_pb2.OffensiveSide.SIDE_MIDDLE, Strategy.SIDE_MIDDLE),
-            (robocup_extension_pb2.OffensiveSide.SIDE_RIGHT, Strategy.SIDE_RIGHT),
-        )
+        self.set_state_defaults()
 
         self.tf_buffer = tf2_ros.Buffer()
         self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self.node)
 
-        # we will try multiple times till we manage to get a connection
-        while rclpy.ok() and self.socket is None:
-            self.socket = self.get_connection()
-            self.node.get_clock().sleep_for(Duration(seconds=1))
-            rclpy.spin_once(self.node)
-        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+        self.try_to_establish_connection()
+        self.run_spin_in_thread()
+
+        self.node.create_timer(1 / self.rate, self.send_message)
+        self.receive_forever()
 
+    def run_spin_in_thread(self):
         # Necessary in ROS2, else we are forever stuck receiving messages
-        thread = threading.Thread(target=rclpy.spin, args=(self.node), daemon=True)
+        thread = threading.Thread(target=rclpy.spin, args=[self.node], daemon=True)
         thread.start()
 
-        self.node.create_timer(1 / self.rate, self.send_message)
-        self.receive_forever()
+    def set_state_defaults(self):
+        self.gamestate: GameState = None
```

Comment:
How does the typing in python work here,
is every field always `Optional` unless it is directly set in the constructor?
Ideally i'd like my types to not be required to be initialized in the constructor, but still ensure that the user will not get the idea to set them to `None`.
Is that at all possible?

Reply:
I would argue that it is Optional if it has the possibility of being None. I don't know if it is possible in the type hint system to set something to None initially, then set a non None value and require non None values afterward while still being mutable. My best guess of enforcing this would be the usage of a setter method that takes the non Optional input, having the internal private variable initialized with None and the Optional type. And access is granted via a getter with the Optional return type, as we could call it before calling the setter.

Is this what you meant?